---
title: "Analysis Report"
author: "Gerda Lukosiute"
output: html_document
---

``` {r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## This is a scRNAseq data analysis of CRC tumor infiltrating cells (heavily focused on NK and ILC1).

``` {r packages, echo = TRUE}
setwd("/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC")
library(Seurat)
library(SeuratDisk)
library(ggplot2)
library(DoubletFinder)
library(tidyverse)
library(SeuratData)
library(reshape2)
library(ggplot2)
library(metap)
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
library(cluster)
library(Nebulosa)
library(dplyr)
library(reshape2)
library(ggrepel)
library(enrichplot)
library(EnhancedVolcano)
library(DESeq2) 
library(readxl)
library(patchwork)
library(tidyr)
library(Matrix)
```

## Initially I exclude sample 5, since it has insufficient data to contribute to the analysis efficiently. I will try to subsequently save it in a later chunck.

``` {r import data, echo = TRUE}
mtx_sample1 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag01_mm/Fionda-2024-R2_SampleTag01_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag01_mm/Fionda-2024-R2_SampleTag01_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag01_mm/Fionda-2024-R2_SampleTag01_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

mtx_sample2 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag02_mm/Fionda-2024-R2_SampleTag02_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag02_mm/Fionda-2024-R2_SampleTag02_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag02_mm/Fionda-2024-R2_SampleTag02_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

mtx_sample3 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag03_mm/Fionda-2024-R2_SampleTag03_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag03_mm/Fionda-2024-R2_SampleTag03_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag03_mm/Fionda-2024-R2_SampleTag03_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

mtx_sample4 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag04_mm/Fionda-2024-R2_SampleTag04_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag04_mm/Fionda-2024-R2_SampleTag04_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag04_mm/Fionda-2024-R2_SampleTag04_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

#mtx_sample5 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag05_mm/Fionda-2024-R2_SampleTag05_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag05_mm/Fionda-2024-R2_SampleTag05_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag05_mm/Fionda-2024-R2_SampleTag05_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

mtx_sample6 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag06_mm/Fionda-2024-R2_SampleTag06_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag06_mm/Fionda-2024-R2_SampleTag06_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag06_mm/Fionda-2024-R2_SampleTag06_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

mtx_sample7 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag07_mm/Fionda-2024-R2_SampleTag07_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag07_mm/Fionda-2024-R2_SampleTag07_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag07_mm/Fionda-2024-R2_SampleTag07_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")

mtx_sample8 <- ReadMtx(mtx = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag08_mm/Fionda-2024-R2_SampleTag08_mm_RSEC_MolsPerCell_MEX/matrix.mtx.gz", features = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag08_mm/Fionda-2024-R2_SampleTag08_mm_RSEC_MolsPerCell_MEX/features.tsv.gz", cells = "/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/mCRC_ILC/Fionda-2024-R2_SampleTag08_mm/Fionda-2024-R2_SampleTag08_mm_RSEC_MolsPerCell_MEX/barcodes.tsv.gz")
```

## Next, I create the Seurat objects for each sample.

``` {r create objects, echo = TRUE}
seurat_s1 <- CreateSeuratObject(counts = mtx_sample1, project = "Sample_1", min.cells = 3, min.features = 200)
seurat_s2 <- CreateSeuratObject(counts = mtx_sample2, project = "Sample_2", min.cells = 3, min.features = 200)
seurat_s3 <- CreateSeuratObject(counts = mtx_sample3, project = "Sample_3", min.cells = 3, min.features = 200)
seurat_s4 <- CreateSeuratObject(counts = mtx_sample4, project = "Sample_4", min.cells = 3, min.features = 200)
#seurat_s5 <- CreateSeuratObject(counts = mtx_sample5, project = "Sample_5", min.cells = 3, min.features = 200)
seurat_s6 <- CreateSeuratObject(counts = mtx_sample6, project = "Sample_6", min.cells = 3, min.features = 200)
seurat_s7 <- CreateSeuratObject(counts = mtx_sample7, project = "Sample_7", min.cells = 3, min.features = 200)
seurat_s8 <- CreateSeuratObject(counts = mtx_sample8, project = "Sample_8", min.cells = 3, min.features = 200)
```

## Subsequently, I perform some of the steps of the Seurat analysis as described in the documentation for Seurat project (https://satijalab.org/seurat/articles/integration_introduction). The remaining steps are completed after merging the samples.

### Sample 1

``` {r sample 1, echo = TRUE}
#Quality control
seurat_s1[["percent.mt"]] <- PercentageFeatureSet(seurat_s1, pattern = "^mt-") 

VlnPlot(seurat_s1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s1, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s1, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s1, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s1.filtered <- subset(seurat_s1, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s1.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s1.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s1.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

#Normalization
s1_normalized <- NormalizeData(object = seurat_s1.filtered)

#Integration
s1_integrated <- FindVariableFeatures(object = s1_normalized)
top10 <- head(VariableFeatures(s1_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s1_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s1_scaled <- ScaleData(object = s1_integrated)

#Linear dimensionality reduction
s1_reduced <- RunPCA(object = s1_scaled) 
VizDimLoadings(s1_reduced, dims = 1:2, reduction = "pca")
DimPlot(s1_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s1_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s1_reduced)

#Clustering
s1_neighboors <- FindNeighbors(object = s1_reduced, dims = 1:20)
s1_clusters <- FindClusters(object = s1_neighboors)
head(Idents(s1_clusters), 5) 

#UMAP
s1_umap <- RunUMAP(object = s1_clusters, dims = 1:20)
DimPlot(s1_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s1 <- paramSweep(s1_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s1 <- summarizeSweep(sweep.res.list_s1, GT = FALSE)
bcmvn_s1 <- find.pK(sweep.stats_s1)

ggplot(bcmvn_s1, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.29 max
pK <- bcmvn_s1[bcmvn_s1$BCmetric == max(bcmvn_s1$BCmetric), "pK"]
pK <- as.numeric(as.character(pK)) 
pK

#Homotypic doublet proportion estimate
annotations <- s1_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s1_umap@meta.data)) ##in project description
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj 

#DoubletFinder
s1_finalized <- doubletFinder(s1_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s1_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.29_67")
table(s1_finalized@meta.data$DF.classifications_0.25_0.29_67)

s1_finalized <- subset(s1_finalized, subset = DF.classifications_0.25_0.29_67 == "Singlet")
table(s1_finalized@meta.data$DF.classifications_0.25_0.29_67)
DimPlot(s1_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.29_67")
```

### Sample 2

``` {r sample 2, echo = TRUE}
#Quality control
seurat_s2[["percent.mt"]] <- PercentageFeatureSet(seurat_s2, pattern = "^mt-") 

VlnPlot(seurat_s2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s2, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s2, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s2.filtered <- subset(seurat_s2, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s2.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s2.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s2.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

#Normalization
s2_normalized <- NormalizeData(object = seurat_s2.filtered)

#Integration
s2_integrated <- FindVariableFeatures(object = s2_normalized)
top10 <- head(VariableFeatures(s2_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s2_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s2_scaled <- ScaleData(object = s2_integrated)

#Linear dimensionality reduction
s2_reduced <- RunPCA(object = s2_scaled) 
VizDimLoadings(s2_reduced, dims = 1:2, reduction = "pca")
DimPlot(s2_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s2_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s2_reduced)

#Clustering
s2_neighboors <- FindNeighbors(object = s2_reduced, dims = 1:20)
s2_clusters <- FindClusters(object = s2_neighboors)
head(Idents(s2_clusters), 5) 

#UMAP
s2_umap <- RunUMAP(object = s2_clusters, dims = 1:20)
DimPlot(s2_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s2 <- paramSweep(s2_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s2 <- summarizeSweep(sweep.res.list_s2, GT = FALSE)
bcmvn_s2 <- find.pK(sweep.stats_s2)

ggplot(bcmvn_s2, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.11 max
pK <- bcmvn_s2[bcmvn_s2$BCmetric == max(bcmvn_s2$BCmetric), "pK"]
pK <- as.numeric(as.character(pK)) 
pK

#Homotypic doublet proportion estimate
annotations <- s2_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s2_umap@meta.data)) ##in project description
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj

#DoubletFinder
s2_finalized <- doubletFinder(s2_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s2_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.11_34")
table(s2_finalized@meta.data$DF.classifications_0.25_0.11_34)

s2_finalized <- subset(s2_finalized, subset = DF.classifications_0.25_0.11_34 == "Singlet")
table(s2_finalized@meta.data$DF.classifications_0.25_0.11_34)
DimPlot(s2_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.11_34")
```

### Sample 3 

``` {r sample 3, echo = TRUE} 
#Quality control
seurat_s3[["percent.mt"]] <- PercentageFeatureSet(seurat_s3, pattern = "^mt-") 

VlnPlot(seurat_s3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s3, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s3, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s3, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s3.filtered <- subset(seurat_s3, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s3.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s3.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s3.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

#Normalization
s3_normalized <- NormalizeData(object = seurat_s3.filtered)

#Integration
s3_integrated <- FindVariableFeatures(object = s3_normalized)
top10 <- head(VariableFeatures(s3_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s3_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s3_scaled <- ScaleData(object = s3_integrated)

#Linear dimensionality reduction
s3_reduced <- RunPCA(object = s3_scaled) 
VizDimLoadings(s3_reduced, dims = 1:2, reduction = "pca")
DimPlot(s3_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s3_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s3_reduced)

#Clustering
s3_neighboors <- FindNeighbors(object = s3_reduced, dims = 1:20)
s3_clusters <- FindClusters(object = s3_neighboors)
head(Idents(s3_clusters), 5) #cluster IDs 1st 5 cells

#UMAP
s3_umap <- RunUMAP(object = s3_clusters, dims = 1:20)
DimPlot(s3_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s3 <- paramSweep(s3_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s3 <- summarizeSweep(sweep.res.list_s3, GT = FALSE)
bcmvn_s3 <- find.pK(sweep.stats_s3)

ggplot(bcmvn_s3, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.25 max
pK <- bcmvn_s3[bcmvn_s3$BCmetric == max(bcmvn_s3$BCmetric), "pK"]
pK <- as.numeric(as.character(pK))
pK

#Homotypic doublet proportion estimate
annotations <- s3_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s3_umap@meta.data))
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj 

#DoubletFinder
s3_finalized <- doubletFinder(s3_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s3_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.25_31")
table(s3_finalized@meta.data$DF.classifications_0.25_0.25_31)

s3_finalized <- subset(s3_finalized, subset = DF.classifications_0.25_0.25_31 == "Singlet")
table(s3_finalized@meta.data$DF.classifications_0.25_0.25_31)
DimPlot(s3_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.25_31")
```

### Sample 4

``` {r sample 4, echo = TRUE}
#Quality control
seurat_s4[["percent.mt"]] <- PercentageFeatureSet(seurat_s4, pattern = "^mt-") 

VlnPlot(seurat_s4, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s4, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s4, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s4, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s4.filtered <- subset(seurat_s4, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s4.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s4.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s4.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

#Normalization
s4_normalized <- NormalizeData(object = seurat_s4.filtered)

#Integration
s4_integrated <- FindVariableFeatures(object = s4_normalized)
top10 <- head(VariableFeatures(s4_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s4_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s4_scaled <- ScaleData(object = s4_integrated)

#Linear dimensionality reduction
s4_reduced <- RunPCA(object = s4_scaled) 
VizDimLoadings(s4_reduced, dims = 1:2, reduction = "pca")
DimPlot(s4_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s4_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s4_reduced)

#Clustering
s4_neighboors <- FindNeighbors(object = s4_reduced, dims = 1:20)
s4_clusters <- FindClusters(object = s4_neighboors)
head(Idents(s4_clusters), 5) 

#UMAP
s4_umap <- RunUMAP(object = s4_clusters, dims = 1:20)
DimPlot(s4_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s4 <- paramSweep(s4_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s4 <- summarizeSweep(sweep.res.list_s4, GT = FALSE)
bcmvn_s4 <- find.pK(sweep.stats_s4)

ggplot(bcmvn_s4, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.3 max
pK <- bcmvn_s4[bcmvn_s4$BCmetric == max(bcmvn_s4$BCmetric), "pK"]
pK <- as.numeric(as.character(pK))
pK

#Homotypic doublet proportion estimate
annotations <- s4_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s4_umap@meta.data)) ##in project description
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj 

#DoubletFinder
s4_finalized <- doubletFinder(s4_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s4_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.3_3")
table(s4_finalized@meta.data$DF.classifications_0.25_0.3_3)

s4_finalized <- subset(s4_finalized, subset = DF.classifications_0.25_0.3_3 == "Singlet")
table(s4_finalized@meta.data$DF.classifications_0.25_0.3_3)
DimPlot(s4_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.3_3")
```

### Sample 5 (excluded))

``` {r sample 5, echo = TRUE}
#Quality control
#seurat_s5[["percent.mt"]] <- PercentageFeatureSet(seurat_s5, pattern = "^mt-") 

#VlnPlot(seurat_s5, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#FeatureScatter(seurat_s5, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
#FeatureScatter(seurat_s5, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
#FeatureScatter(seurat_s5, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

#seurat_s5.filtered <- subset(seurat_s5, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10) #maybe lower percent.mt?

#FeatureScatter(seurat_s5.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
#FeatureScatter(seurat_s5.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
#FeatureScatter(seurat_s5.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

#VlnPlot(seurat_s5.filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#Normalization
#s5_normalized <- NormalizeData(object = seurat_s5.filtered)

#Integration
#s5_integrated <- FindVariableFeatures(object = s5_normalized)
#top10 <- head(VariableFeatures(s5_integrated), 10)
#top10
#plot1 <- VariableFeaturePlot(s5_integrated)
#plot1
#plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
# plot2
# 
# #Scaling
# s5_scaled <- ScaleData(object = s5_integrated)
# 
# #Linear dimensionality reduction
# s5_reduced <- RunPCA(object = s5_scaled, npcs = 10) 
# VizDimLoadings(s5_reduced, dims = 1:2, reduction = "pca")
# DimPlot(s5_reduced, reduction = "pca") + NoLegend()
# DimHeatmap(s5_reduced, dims = 1:10, cells = 500, balanced = TRUE)
# ElbowPlot(s5_reduced)
# 
# #Clustering
# s5_neighboors <- FindNeighbors(object = s5_reduced, dims = 1:10)
# s5_clusters <- FindClusters(object = s5_neighboors)
# head(Idents(s5_clusters), 5) 
# 
# #UMAP
# s5_umap <- RunUMAP(object = s5_clusters, dims = 1:10, n.neighbors = 5)
# DimPlot(s5_umap, reduction = "umap")
# 
# #Doublet finder; UNAVAILABLE FOR THIS SAMPLE; RUN DOWNSTREAM ANALYSIS WITH AND WITHOUT S5 TO SEE THE EFFECT
# #pK identification (no ground-truth)
# sweep.res.list_s5 <- paramSweep(s5_umap, PCs = 1:10, sct = FALSE) #very low cell count, gives error 
# sweep.stats_s5 <- summarizeSweep(sweep.res.list_s5, GT = FALSE)
# bcmvn_s5 <- find.pK(sweep.stats_s5)
# 
# ggplot(bcmvn_s5, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
# #0.3 max
# pK <- bcmvn_s5[bcmvn_s4$BCmetric == max(bcmvn_s5$BCmetric), "pK"]
# pK <- as.numeric(as.character(pK))
# pK
# 
# #Homotypic doublet proportion estimate
# annotations <- s5_umap@meta.data$seurat_clusters
# homotypic.prop <- modelHomotypic(annotations)
# nExp_poi <- round(0.05*nrow(s5_umap@meta.data)) 
# nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
# nExp_poi.adj 
# 
# #DoubletFinder
# s5_finalized <- doubletFinder(s5_umap,
#                               PCs = 1:20,
#                               pK = pK,
#                               nExp = nExp_poi.adj,
#                               reuse.pANN = FALSE, sct = FALSE)
# DimPlot(s5_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.3_3")
# table(s5_finalized@meta.data$DF.classifications_0.25_0.3_3)
# 
# s5_finalized <- subset(s5_finalized, subset = DF.classifications_0.25_0.3_3 == "Singlet")
# table(s5_finalized@meta.data$DF.classifications_0.25_0.3_3)
# DimPlot(s5_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.3_3")
```

### Sample 6

``` {r sample 6, echo = TRUE}
#Quality control
seurat_s6[["percent.mt"]] <- PercentageFeatureSet(seurat_s6, pattern = "^mt-") 

VlnPlot(seurat_s6, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s6, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s6, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s6, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s6.filtered <- subset(seurat_s6, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s6.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s6.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s6.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

VlnPlot(seurat_s6.filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#Normalization
s6_normalized <- NormalizeData(object = seurat_s6.filtered)

#Integration
s6_integrated <- FindVariableFeatures(object = s6_normalized)
top10 <- head(VariableFeatures(s6_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s6_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s6_scaled <- ScaleData(object = s6_integrated)

#Linear dimensionality reduction
s6_reduced <- RunPCA(object = s6_scaled) 
VizDimLoadings(s6_reduced, dims = 1:2, reduction = "pca")
DimPlot(s6_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s6_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s6_reduced)

#Clustering
s6_neighboors <- FindNeighbors(object = s6_reduced, dims = 1:20)
s6_clusters <- FindClusters(object = s6_neighboors)
head(Idents(s6_clusters), 5) 

#UMAP
s6_umap <- RunUMAP(object = s6_clusters, dims = 1:20)
DimPlot(s6_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s6 <- paramSweep(s6_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s6 <- summarizeSweep(sweep.res.list_s6, GT = FALSE)
bcmvn_s6 <- find.pK(sweep.stats_s6)

ggplot(bcmvn_s6, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.2
pK <- bcmvn_s6[bcmvn_s6$BCmetric == max(bcmvn_s6$BCmetric), "pK"]
pK <- as.numeric(as.character(pK))
pK

#Homotypic doublet proportion estimate
annotations <- s6_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s6_umap@meta.data)) ##in project description
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj 

#DoubletFinder
s6_finalized <- doubletFinder(s6_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s6_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.2_37")
table(s6_finalized@meta.data$DF.classifications_0.25_0.2_37)

s6_finalized <- subset(s6_finalized, subset = DF.classifications_0.25_0.2_37 == "Singlet")
table(s6_finalized@meta.data$DF.classifications_0.25_0.2_37)
DimPlot(s6_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.2_37")
```

### Sample 7

``` {r sample 7, echo = TRUE}
#Quality control
seurat_s7[["percent.mt"]] <- PercentageFeatureSet(seurat_s7, pattern = "^mt-") 

VlnPlot(seurat_s7, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s7, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s7, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s7, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s7.filtered <- subset(seurat_s7, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s7.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s7.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s7.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

VlnPlot(seurat_s7.filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#Normalization
s7_normalized <- NormalizeData(object = seurat_s7.filtered)

#Integration
s7_integrated <- FindVariableFeatures(object = s7_normalized)
top10 <- head(VariableFeatures(s7_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s7_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s7_scaled <- ScaleData(object = s7_integrated)

#Linear dimensionality reduction
s7_reduced <- RunPCA(object = s7_scaled) 
VizDimLoadings(s7_reduced, dims = 1:2, reduction = "pca")
DimPlot(s7_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s7_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s7_reduced)

#Clustering
s7_neighboors <- FindNeighbors(object = s7_reduced, dims = 1:20)
s7_clusters <- FindClusters(object = s7_neighboors)
head(Idents(s7_clusters), 5) 

#UMAP
s7_umap <- RunUMAP(object = s7_clusters, dims = 1:20)
DimPlot(s7_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s7 <- paramSweep(s7_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s7 <- summarizeSweep(sweep.res.list_s7, GT = FALSE)
bcmvn_s7 <- find.pK(sweep.stats_s7)

ggplot(bcmvn_s7, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.16
pK <- bcmvn_s7[bcmvn_s7$BCmetric == max(bcmvn_s7$BCmetric), "pK"]
pK <- as.numeric(as.character(pK))
pK

#Homotypic doublet proportion estimate
annotations <- s7_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s7_umap@meta.data)) ##in project description
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj 

#DoubletFinder
s7_finalized <- doubletFinder(s7_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s7_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.16_9")
table(s7_finalized@meta.data$DF.classifications_0.25_0.16_9)

s7_finalized <- subset(s7_finalized, subset = DF.classifications_0.25_0.16_9 == "Singlet")
table(s7_finalized@meta.data$DF.classifications_0.25_0.16_9)
DimPlot(s7_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.16_9")
```

### Sample 8

``` {r sample 8, echo = TRUE}
#Quality control
seurat_s8[["percent.mt"]] <- PercentageFeatureSet(seurat_s8, pattern = "^mt-") 

VlnPlot(seurat_s8, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FeatureScatter(seurat_s8, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s8, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s8, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 

seurat_s8.filtered <- subset(seurat_s8, subset = nFeature_RNA >= 200 & nFeature_RNA <= 4000 & percent.mt < 10)

FeatureScatter(seurat_s8.filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
FeatureScatter(seurat_s8.filtered, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm') 
FeatureScatter(seurat_s8.filtered, feature1 = "nFeature_RNA", feature2 = "percent.mt") + geom_smooth(method = 'lm')

VlnPlot(seurat_s8.filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#Normalization
s8_normalized <- NormalizeData(object = seurat_s8.filtered)

#VFs
s8_integrated <- FindVariableFeatures(object = s8_normalized)
top10 <- head(VariableFeatures(s8_integrated), 10)
top10
plot1 <- VariableFeaturePlot(s8_integrated)
plot1
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) 
plot2

#Scaling
s8_scaled <- ScaleData(object = s8_integrated)

#Linear dimensionality reduction
s8_reduced <- RunPCA(object = s8_scaled) 
VizDimLoadings(s8_reduced, dims = 1:2, reduction = "pca")
DimPlot(s8_reduced, reduction = "pca") + NoLegend()
DimHeatmap(s8_reduced, dims = 1:20, cells = 500, balanced = TRUE)
ElbowPlot(s8_reduced)

#Clustering
s8_neighboors <- FindNeighbors(object = s8_reduced, dims = 1:20)
s8_clusters <- FindClusters(object = s8_neighboors)
head(Idents(s8_clusters), 5) #cluster IDs 1st 5 cells

#UMAP
s8_umap <- RunUMAP(object = s8_clusters, dims = 1:20)
DimPlot(s8_umap, reduction = "umap")

#Doublet finder
#pK identification (no ground-truth)
sweep.res.list_s8 <- paramSweep(s8_umap, PCs = 1:20, sct = FALSE)
sweep.stats_s8 <- summarizeSweep(sweep.res.list_s8, GT = FALSE)
bcmvn_s8 <- find.pK(sweep.stats_s8)

ggplot(bcmvn_s8, aes(pK, BCmetric, group = 1)) + geom_point() + geom_line()
#0.15
pK <- bcmvn_s8[bcmvn_s8$BCmetric == max(bcmvn_s8$BCmetric), "pK"]
pK <- as.numeric(as.character(pK))
pK

#Homotypic doublet proportion estimate
annotations <- s8_umap@meta.data$seurat_clusters
homotypic.prop <- modelHomotypic(annotations)
nExp_poi <- round(0.05*nrow(s8_umap@meta.data)) ##in project description
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
nExp_poi.adj 

#DoubletFinder
s8_finalized <- doubletFinder(s8_umap,
                              PCs = 1:20,
                              pK = pK,
                              nExp = nExp_poi.adj,
                              reuse.pANN = FALSE, sct = FALSE)
DimPlot(s8_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.15_24")
table(s8_finalized@meta.data$DF.classifications_0.25_0.15_24)

s8_finalized <- subset(s8_finalized, subset = DF.classifications_0.25_0.15_24 == "Singlet")
table(s8_finalized@meta.data$DF.classifications_0.25_0.15_24)
DimPlot(s8_finalized, reduction = 'umap', group.by = "DF.classifications_0.25_0.15_24")
```

## Next, merge the samples into a single object for easier downstream analysis (skipping sample 5).

``` {r merge object, echo = TRUE}
samples <- list(s2_finalized, s3_finalized, s4_finalized, s6_finalized, s7_finalized, s8_finalized) #skipping s5
merged_object <- merge(s1_finalized, y = samples, add.cell.ids = c("s1", "s2", "s3", "s4", "s6", "s7", "s8"))
```

## Proceeding with the steps in the Seurat documentation on integration.

``` {r merged downstream, echo = TRUE} 
merged_object <- NormalizeData(merged_object)
merged_object <- FindVariableFeatures(merged_object)
merged_object <- ScaleData(merged_object)
merged_object <- RunPCA(merged_object)
merged_object <- FindNeighbors(merged_object, dims = 1:20, reduction = "pca")
merged_object <- FindClusters(merged_object, resolution = 2, cluster.name = "unintegrated_clusters")
merged_object <- RunUMAP(merged_object, dims = 1:20, reduction = "pca", reduction.name = "umap.unintegrated")
DimPlot(merged_object, reduction ="umap.unintegrated", group.by = c("orig.ident", "seurat_clusters"))
merged_object <- IntegrateLayers(object = merged_object, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca", verbose = FALSE)

merged_object <- JoinLayers(merged_object)
head(merged_object)
tail(merged_object)

merged_object <- FindNeighbors(merged_object, reduction = "integrated.cca", dims = 1:20)
merged_object <- FindClusters(merged_object, resolution = 0.5, cluster.name = "integrated_clusters") 
merged_object <- RunUMAP(merged_object, dims = 1:20, reduction = "integrated.cca")
DimPlot(merged_object, reduction = "umap", group.by = c("orig.ident", "seurat_clusters")) 
DimPlot(merged_object, reduction = "umap", split.by = "orig.ident")
```

## Percentage of samples in each cluster.

``` {r percent per cluster, echo = TRUE}
cell_counts <- table(merged_object$orig.ident, merged_object$integrated_clusters)
cell_perc <- prop.table(cell_counts, margin = 1)*100
percentages_df <- melt(cell_perc, varnames = c("Sample", "Cluster"), value.name = "Percentage")

ggplot(percentages_df, aes(x = as.factor(Cluster), y = Percentage, fill = Sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Percentage of cells per sample in each cluster",
       x = "Cluster",
       y = "Percentage of Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

## Finding all markers

``` {r type annotation, echo = TRUE}
all_markers <- FindAllMarkers(merged_object, only.pos = T)
top_markers <- all_markers %>%
  group_by(cluster) %>%
  slice_max(order_by = avg_log2FC, n = 20)
print(top_markers, n = 200)
```

## Conserved markers by clusters

### Cluster 0

``` {r cluster 0, echo = TRUE}
markers_c0 <- FindConservedMarkers(merged_object, ident.1 = 0, grouping.var = "orig.ident")
head(markers_c0)
FeaturePlot(merged_object, features = c("Gzma", "Gzmb", "Tmsb4x", "Prf1", "Actb", "Emb"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 1

``` {r cluster 1, echo = TRUE} 
markers_c1 <- FindConservedMarkers(merged_object, ident.1 = 1, grouping.var = "orig.ident")
head(markers_c1)
FeaturePlot(merged_object, features = c("Ccr2", "Ccl5", "Gzmb", "Cd7", "Rap1b", "Ctla2a"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 2 

``` {r cluster 2, echo = TRUE}
markers_c2 <- FindConservedMarkers(merged_object, ident.1 = 2, grouping.var = "orig.ident")
head(markers_c2)
FeaturePlot(merged_object, features = c("Tmsb4x", "Pmepa1", "Cfl1", "Actb", "Ncr1", "Pfn1"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 3

``` {r cluster 3, echo = TRUE}
markers_c3 <- FindConservedMarkers(merged_object, ident.1 = 3, grouping.var = "orig.ident")
head(markers_c3)
FeaturePlot(merged_object, features = c("Nr4a1", "Pim1", "Nr4a3", "Nfkbiz", "Nfkbid", "Nfkbia"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 4

``` {r cluster 4, echo = TRUE}
markers_c4 <- FindConservedMarkers(merged_object, ident.1 = 4, grouping.var = "orig.ident")
head(markers_c4)
FeaturePlot(merged_object, features = c("Tmem176b", "Tmem176a", "Cxcr6", "Il7r", "Ckb", "Krt83"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 5

``` {r cluster 5, echo = TRUE}
markers_c5 <- FindConservedMarkers(merged_object, ident.1 = 5, grouping.var = "orig.ident")
head(markers_c5)
FeaturePlot(merged_object, features = c("Cd3e", "Cd3g", "Trgc2", "Cd3d", "Trgc1", "Dpp4"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 6

``` {r cluster 6, echo = TRUE}
markers_c6 <- FindConservedMarkers(merged_object, ident.1 = 6, grouping.var = "orig.ident")
head(markers_c6)
FeaturePlot(merged_object, features = c("Oas3", "Mx1", "Rsad2", "Ifit1", "Slfn5", "Ifi204"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 7

``` {r cluster 7, echo = TRUE}
markers_c7 <- FindConservedMarkers(merged_object, ident.1 = 7, grouping.var = "orig.ident")
head(markers_c7)
FeaturePlot(merged_object, features = c("Cd8b1", "Cd8a", "Themis", "Pdcd1", "Fam169b", "Havcr2"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 8

``` {r cluster 8, echo = TRUE}
markers_c8 <- FindConservedMarkers(merged_object, ident.1 = 8, grouping.var = "orig.ident")
head(markers_c8)
FeaturePlot(merged_object, features = c("Pclaf", "E2f1", "Kntc1", "Nuf2", "Spc24", "Cdc6"), min.cutoff = 'q10', reduction = "umap")
```

### Cluster 9

``` {r cluster 9, echo = TRUE}
markers_c9 <- FindConservedMarkers(merged_object, ident.1 = 9, grouping.var = "orig.ident")
head(markers_c9)
FeaturePlot(merged_object, features = c("Cd40", "Clec4d", "Sirpa", "Pirb", "Tgfbi", "Il1b"), min.cutoff = 'q10', reduction = "umap")
```

# Cell typist annotation (refer to the celltypist_run.py script in the repo) 

``` {r annotation, echo = TRUE}
expression_matrix <- GetAssayData(merged_object, assay = "RNA", slot = "data")
dense_matrix <- as.matrix(expression_matrix)
expression_df <- as.data.frame(t(dense_matrix))
expression_df$cell_id <- rownames(expression_df)
expression_df <- expression_df[, c("cell_id", setdiff(colnames(expression_df), "cell_id"))]
write.csv(expression_df, file = "seurat_expression.csv", row.names = FALSE, quote = FALSE)

predictions <- read.csv("/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/predicted_labels.csv")
head(predictions)
colnames(predictions)[1] <- "cell_id"
rownames(predictions) <- predictions$cell_id
predictions <- predictions[colnames(merged_object), ]  
merged_object <- AddMetaData(merged_object, metadata = predictions$predicted_label, col.name = "celltypist_labels")
table(merged_object@meta.data$celltypist_labels)


DimPlot(merged_object,
        reduction = "umap",
        group.by = c("celltypist_labels", "seurat_clusters"),
        label = FALSE)

label_counts <- as.data.frame(table(merged_object@meta.data$celltypist_labels))
colnames(label_counts) <- c("Cell Type", "Cell Count")

# table
ggplot(label_counts, aes(x = 1, y = reorder(`Cell Type`, -`Cell Count`))) +
  geom_text(aes(label = paste0(`Cell Type`, ": ", `Cell Count`)), hjust = 0) +
  theme_void() +
  theme(
    plot.margin = margin(1, 1, 1, 1, "cm"),
    text = element_text(size = 12),
    panel.grid = element_blank()
  ) +
  xlim(1, 1.5) 

label_counts$`Cell Type` <- factor(label_counts$`Cell Type`, levels = label_counts$`Cell Type`[order(label_counts$`Cell Count`, decreasing = TRUE)])

header_row <- data.frame(
  `Cell Type` = "Cell Type",
  `Cell Count` = "Count",
  check.names = FALSE
)

table_data <- rbind(header_row, label_counts)
table_data$y_pos <- rev(seq_len(nrow(table_data))) 

p3 <- ggplot(table_data, aes(y = y_pos)) +
  geom_text(aes(x = 0, label = `Cell Type`), hjust = 0, size = 4, fontface = ifelse(table_data$y_pos == max(table_data$y_pos), "bold", "plain")) +
  geom_text(aes(x = 1, label = `Cell Count`), hjust = 1, size = 4, fontface = ifelse(table_data$y_pos == max(table_data$y_pos), "bold", "plain")) +
  geom_segment(aes(x = -0.05, xend = 1.05, y = max(y_pos) - 0.5, yend = max(y_pos) - 0.5), size = 0.5) + # Header line
  theme_void() +
  theme(
    plot.margin = margin(1, 1, 1, 1, "cm"),
    plot.title = element_text(hjust = 0.5, face = "bold", margin = margin(b = 10)),
    panel.grid = element_blank(),
    axis.title.y = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank()
  ) +
  xlim(-0.1, 1.1) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) 

print(p3)

```

# Initial NK and ILC1 subset

``` {r subset1, echo = TRUE}
FeaturePlot(merged_object, features = c("Cd3e", "Cd3d", "Tbx21", "Eomes", "Cd8b1", "Nkg7", "Gzmb"), min.cutoff = 'q10', reduction = "umap")

NK_and_ILC1 <- subset(merged_object, seurat_clusters %in% c(0, 1, 2, 3, 4, 6)) #both, 4717
NK_and_ILC1 <- NormalizeData(NK_and_ILC1)
NK_and_ILC1 <- FindVariableFeatures(NK_and_ILC1)
NK_and_ILC1 <- ScaleData(NK_and_ILC1, features = rownames(NK_and_ILC1))
NK_and_ILC1 <- RunPCA(NK_and_ILC1)
NK_and_ILC1 <- FindNeighbors(NK_and_ILC1, dims = 1:20)
NK_and_ILC1 <- FindClusters(NK_and_ILC1, resolution = 0.5, cluster.name = "NK_ILC1_clusters") 
NK_and_ILC1 <- RunUMAP(NK_and_ILC1, dims = 1:20)
DimPlot(NK_and_ILC1, reduction = "umap", group.by = "NK_ILC1_clusters")

```

# Assigning m95/96 organoids
## m95 -> 4 mutations (Apc, Kras, P53, Smad4)
## m96 -> 3 mutations (Apc, Kras, P53)

```{r organoids, echo=TRUE} 
# organoids
NK_and_ILC1$organoid <- ifelse(
  grepl("^s[1-4]_", colnames(NK_and_ILC1)), "m95", 
  ifelse(grepl("^s[6-8]_", colnames(NK_and_ILC1)), "m96", NA)
)

simple_shared <- table(NK_and_ILC1$organoid)
simple_shared

Idents(NK_and_ILC1) <- "organoid" 

DimPlot(NK_and_ILC1, reduction = "umap", split.by = "organoid", group.by = "NK_ILC1_clusters") + ggtitle("NK and ILC1 Clusters Split by Organoid")

shared_table <- table(NK_and_ILC1$NK_ILC1_clusters, NK_and_ILC1$organoid)
shared_table
m95_col <- shared_table[,"m95"] 
m96_col <- shared_table[,"m96"] 

m_95 <- data.frame(m95_col)
m95 <- rownames(m_95)
ggplot(m_95, aes(x="", y=m_95[,1], fill = m95)) + geom_bar(stat = "identity", width = 1, color = "white") + coord_polar("y", start = 0) + theme_void() 

m_96 <- data.frame(m96_col)
m96 <- rownames(m_96)
ggplot(m_96, aes(x="", y=m_96[,1], fill = m96)) + geom_bar(stat = "identity", width = 1, color = "white") + coord_polar("y", start = 0) + theme_void() 

counts_df <- melt(shared_table, varnames = c("Cluster", "Organoid"), value.name = "Count")

ggplot(counts_df, aes(x = as.factor(Cluster), y = Count, fill = Organoid)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Cell Count per Cluster in Each Organoid",
       x = "Cluster",
       y = "Cell Count") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# STATISTICAL SIGNIFICANCE
# tests cluster count identity differences between samples 
chisq.test(shared_table)  

# for small sample sizes
fisher.test(shared_table, simulate.p.value = TRUE, B = 10000)

# each cluster individually using proportion test
for (i in 1:nrow(shared_table)) {
  cat("Cluster", rownames(shared_table)[i], "\n")
  print(prop.test(shared_table[i,], colSums(shared_table)))
}
# 0 -> < 0.05, but rlly close; likely more represented in m95
# 1 -> < 0.05, also likely more represented in m95
# 2 -> insignificant
# 3 -> < 0.05, also rlly close; significantly over-represented in m95
# 4 -> < 0.05, significantly over-represented in m96
# 5 -> insignificant

# per sample (not per organoid) comparison; avoids pseudoreplication
cluster_props <- NK_and_ILC1@meta.data %>%
  group_by(orig.ident, NK_ILC1_clusters) %>%
  summarise(count = n()) %>%
  group_by(orig.ident) %>%
  mutate(prop = count / sum(count))
cluster_props

# VISUALIZATION
cluster_props$organoid <- case_when(
  cluster_props$orig.ident %in% c("Sample_1", "Sample_2", "Sample_3", "Sample_4") ~ "m95",
  cluster_props$orig.ident %in% c("Sample_6", "Sample_7", "Sample_8") ~ "m96",
  TRUE ~ NA_character_
)

# proportion boxplot (old)
ggplot(cluster_props, aes(x = NK_ILC1_clusters, y = prop, fill = organoid)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color = organoid), width = 0.2, size = 1.5, alpha = 0.7) +
  labs(title = "Cluster Proportions per Sample by Organoid",
       x = "Cluster",
       y = "Proportion") +
  theme_minimal()

# extract actual p-values per cluster from your loop and manually format
pvals_df <- data.frame(
  Cluster = as.factor(0:5),
  pval = c(0.04384, 1.152e-05, 0.3267, 0.03869, 2.2e-16, 0.07369)
)

pvals_df$p_text <- paste0("p = ", signif(pvals_df$pval, 3))

box_and_p <- ggplot(cluster_props, aes(x = as.factor(NK_ILC1_clusters), y = prop, fill = organoid)) +
  geom_boxplot(
    aes(group = interaction(NK_ILC1_clusters, organoid)),
    position = position_dodge(width = 0.6),
    outlier.shape = NA,
    width = 0.5
  ) +
  geom_jitter(
    aes(color = organoid),
    position = position_jitterdodge(dodge.width = 0.6, jitter.width = 0.2),
    size = 2,
    alpha = 0.8
  ) +
  geom_text(data = pvals_df, 
            aes(x = Cluster, y = 0.52, label = p_text), 
            inherit.aes = FALSE, 
            size = 3.8) +
  scale_fill_manual(values = c("m95" = "lightpink", "m96" = "lightblue")) +
  scale_color_manual(values = c("m95" = "red3", "m96" = "blue3")) +
  labs(
    title = "Cluster Proportions per Sample by Organoid",
    x = "Cluster",
    y = "Proportion"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    legend.position = "top",
    panel.grid.major.x = element_blank()
  )

box_and_p
ggsave("cluster_proportions.png", plot = box_and_p, width = 8, height = 6, dpi = 600)

heat_df <- dcast(cluster_props, orig.ident + organoid ~ NK_ILC1_clusters, value.var = "prop")

heat_df[is.na(heat_df)] <- 0 
rownames(heat_df) <- heat_df$orig.ident
heat_data <- as.matrix(heat_df[, -(1:2)])  

# annotation with matching rownames
annotation_row <- data.frame(Organoid = heat_df$organoid)
rownames(annotation_row) <- heat_df$orig.ident  

pheatmap::pheatmap(heat_data, 
                   cluster_rows = TRUE, 
                   cluster_cols = TRUE,
                   annotation_row = annotation_row,
                   main = "Cluster Proportion Heatmap")

# normalized cluster props
prop_df <- as.data.frame(prop.table(shared_table, margin = 2)) 
colnames(prop_df) <- c("Cluster", "Organoid", "Proportion")

ggplot(prop_df, aes(x = Cluster, y = Proportion, fill = Organoid)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Normalized Cluster Proportions by Organoid",
       y = "Proportion", x = "Cluster") +
  theme_minimal()

```

# Maturation/Exhaustion

```{r DEA, echo = TRUE}
# general subset markers wrt other cells; add organoid; fc <= 0.2; DESeq2
Idents(NK_and_ILC1) <- "NK_ILC1_clusters"

# Differential expression using DESeq2 and accounting for organoid as latent variable
#NK_and_ILC1_markers <- FindAllMarkers(
  #object = NK_and_ILC1,
  #test.use = "DESeq2",
  #min.pct = 0.25,
  #logfc.threshold = 0.2,
  #group.by = "NK_ILC1_clusters",
  #latent.vars = "organoid"
#)

head(NK_and_ILC1_markers)

# sub
NK_and_ILC1_markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 0.2) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10pos

NK_and_ILC1_markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC < -0.2) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10neg

DoHeatmap(NK_and_ILC1, features = top10pos$gene, group.by = "NK_ILC1_clusters")
DoHeatmap(NK_and_ILC1, features = top10neg$gene, group.by = "NK_ILC1_clusters")

# maturation
maturation_genes <- c("Itgam", "Klrg1", "Cd27")
NK_and_ILC1$organoid <- factor(NK_and_ILC1$organoid, levels = c("m96", "m95"))

vln_colors <- c("m96" = "lightblue", "m95" = "pink")

FeaturePlot(NK_and_ILC1, features = maturation_genes, min.cutoff = 'q10', reduction = "umap", split.by = "organoid")
VlnPlot(NK_and_ILC1, features = maturation_genes, group.by = "NK_ILC1_clusters", split.by = "organoid", pt.size = 0, cols = c("lightblue", "pink")) 
levels(NK_and_ILC1$organoid)

DotPlot(NK_and_ILC1, features = maturation_genes, group.by = "NK_ILC1_clusters", split.by = "organoid", cols = c("pink", "lightblue")) + RotatedAxis()
RidgePlot(NK_and_ILC1, features = maturation_genes, ncol = 3, group.by = "NK_ILC1_clusters") 
DoHeatmap(NK_and_ILC1, features = maturation_genes, size = 3) 
FeatureScatter(NK_and_ILC1, feature1 = "Klrg1", feature2 = "Cd27")
FeatureScatter(NK_and_ILC1, feature1 = "Itgam", feature2 = "Cd27")

# exhaustion
exhaustion_genes1 <- c("Tcf7", "Pdcd1", "Tox", "Tigit", "Havcr2", "Slamf6", "Cxcr5", "Nfatc1", "Nfatc2") 
exhaustion_genes2 <- c("Nfatc3", "Nfatc4", "Nfat5", "Nr4a1", "Nr4a2", "Nr4a3", "Lag3", "Ctla4", "Prdm1")  
exhaustion_genes <- c("Tcf7", "Pdcd1", "Tox", "Tigit", "Havcr2", "Slamf6", "Cxcr5", "Nfatc1", "Nfatc2", "Nfatc3", "Nfat5", "Nr4a1", "Nr4a2", "Nr4a3", "Lag3", "Ctla4", "Prdm1")

FeaturePlot(NK_and_ILC1, features = exhaustion_genes1, min.cutoff = 'q10', reduction = "umap", split.by = "NK_ILC1_clusters")
VlnPlot(NK_and_ILC1, features = exhaustion_genes1, group.by = "NK_ILC1_clusters", pt.size = 0)
FeaturePlot(NK_and_ILC1, features = exhaustion_genes2, min.cutoff = 'q10', reduction = "umap", split.by = "NK_ILC1_clusters")
VlnPlot(NK_and_ILC1, features = exhaustion_genes2, group.by = "NK_ILC1_clusters", pt.size = 0)


features <- c("Tcf7", "Tox", "Tigit", "Nfatc1", "Nfatc3", "Nfat5", "Nr4a1", "Nr4a2", "Nr4a3", "Ctla4", "Prdm1")

vln_list <- lapply(features, function(gene) {
  VlnPlot(
    object = NK_and_ILC1,
    features = gene,
    group.by = "NK_ILC1_clusters",
    pt.size = 0,
    combine = FALSE
  )[[1]] +
    ggtitle(gene) +
    theme_minimal(base_size = 10) +
    theme(
      axis.title.x = element_blank(),
      #axis.title.y = element_blank(),  # Suppress individual y labels
      plot.title = element_text(face = "bold", hjust = 0.5, size = 10),
      axis.text.x = element_text(angle = 45, hjust = 1, size = 7),
      axis.text.y = element_text(size = 7),
      legend.position = "none"
    )
})

wrap_plots(vln_list, ncol = 4)  

vln_grid <- wrap_plots(vln_list, ncol = 4)
ggsave("updated_06_24/exhaustion_violin_grid.png", plot = vln_grid, width = 10, height = 6, dpi = 600)


DotPlot(NK_and_ILC1, features = exhaustion_genes1, group.by = "NK_ILC1_clusters", split.by = "organoid", cols = c("lightblue", "pink")) + RotatedAxis()

DotPlot(NK_and_ILC1, features = exhaustion_genes2, group.by = "NK_ILC1_clusters", split.by = "organoid", cols = c("lightblue", "pink")) + RotatedAxis()

DotPlot(NK_and_ILC1, features = exhaustion_genes, group.by = "NK_ILC1_clusters", split.by = "organoid", cols = c("lightblue", "pink")) + RotatedAxis()

NK_and_ILC1 <- AddModuleScore(
  NK_and_ILC1,
  exhaustion_genes,
  pool = NULL,
  nbin = 24,
  ctrl = 100,
  k = FALSE,
  assay = NULL,
  name = "Exhaustion_sig",
  seed = 1,
  search = FALSE,
  slot = "data"
)
VlnPlot(NK_and_ILC1, features = "Exhaustion_sig1", split.by = "organoid", group.by = "NK_ILC1_clusters")
```

# cNK vs trNK

```{r c vs tr, echo = TRUE}
# import gene lists
gene_sets <- read_excel("/Users/gerdalukosiute/Downloads/Thesis/analysis_mCRC/circ_vs_TR_genes.xlsx", sheet = 1, skip = 1)
colnames(gene_sets) <- make.names(colnames(gene_sets))

# extract gene lists
residency_genes     <- na.omit(gene_sets$Residency.score..Milner.et.al.) %>% as.character()
circulating_genes   <- na.omit(gene_sets$Circulating.score..Milner.et.al.) %>% as.character()
ILC1_genes          <- na.omit(gene_sets$ILC1.score..Lopes.et.al.) %>% as.character()
cNK_genes           <- na.omit(gene_sets$cNK.score..Lopes.et.al.) %>% as.character()

# module scores
NK_and_ILC1 <- AddModuleScore(NK_and_ILC1, features = list(residency_genes), name = "Residency")
NK_and_ILC1 <- AddModuleScore(NK_and_ILC1, features = list(circulating_genes), name = "Circulating")
NK_and_ILC1 <- AddModuleScore(NK_and_ILC1, features = list(ILC1_genes), name = "ILC1")
NK_and_ILC1 <- AddModuleScore(NK_and_ILC1, features = list(cNK_genes), name = "cNK")

# plot c("Residency1", "Circulating1", "ILC11", "cNK1", "CD56dim1", "CD56bright1")
VlnPlot(NK_and_ILC1, features = "ILC11", split.by = "organoid", group.by = "NK_ILC1_clusters", pt.size = 0.1)
VlnPlot(NK_and_ILC1, features = "cNK1", split.by = "organoid", group.by = "NK_ILC1_clusters", pt.size = 0.1)
VlnPlot(NK_and_ILC1, features = "Circulating1", split.by = "organoid", group.by = "NK_ILC1_clusters", pt.size = 0.1)
VlnPlot(NK_and_ILC1, features = "Residency1", split.by = "organoid", group.by = "NK_ILC1_clusters", pt.size = 0.1)

# features
score_features1 <- c("ILC11", "cNK1")
score_features2 <- c("Circulating1", "Residency1")

# plots with custom median
custom_score_plots1 <- lapply(score_features1, function(feature) {
  midpoint_val <- median(NK_and_ILC1@meta.data[[feature]], na.rm = TRUE)
  
  FeaturePlot(NK_and_ILC1, features = feature, reduction = "umap", slot = "data") +
    scale_color_gradient2(
      low = "purple",
      mid = "white",
      high = "red",
      midpoint = midpoint_val
    ) +
    ggtitle(paste0(feature, " (mid=", round(midpoint_val, 3), ")")) +
    theme_minimal()
})

# display
wrap_plots(custom_score_plots1)

custom_score_plots2 <- lapply(score_features2, function(feature) {
  midpoint_val <- median(NK_and_ILC1@meta.data[[feature]], na.rm = TRUE)
  
  FeaturePlot(NK_and_ILC1, features = feature, reduction = "umap", slot = "data") +
    scale_color_gradient2(
      low = "purple",
      mid = "white",
      high = "red",
      midpoint = midpoint_val
    ) +
    ggtitle(paste0(feature, " (mid=", round(midpoint_val, 3), ")")) +
    theme_minimal()
})

# display
wrap_plots(custom_score_plots2)

combined_plot1 <- wrap_plots(custom_score_plots1, ncol = 2)
combined_plot2 <- wrap_plots(custom_score_plots2, ncol = 2)

# save them
ggsave("updated_06_24/ILC1_cNK_scores_umap.png", plot = combined_plot1, dpi = 600, width = 8, height = 4)
ggsave("updated_06_24/Circulating_Residency_scores_umap.png", plot = combined_plot2, dpi = 600, width = 8, height = 4)

```

# Volcano (3+0 vs. others)

```{r comp, echo = TRUE}
Idents(NK_and_ILC1) <- "NK_ILC1_clusters"
# exclude ILC1s
NK_and_ILC1 <- subset(NK_and_ILC1, NK_and_ILC1$NK_ILC1_clusters %in% c(0, 1, 2, 3, 5))

# define group
NK_and_ILC1$pseudobulk_group <- paste0(
  ifelse(NK_and_ILC1$NK_ILC1_clusters %in% c(0, 3), "0_3", "others"), "_",
  NK_and_ILC1$organoid
)

counts <- GetAssayData(NK_and_ILC1, slot = "counts")
meta <- NK_and_ILC1@meta.data

# sum counts by group
pseudobulk_counts <- tapply(colnames(counts), meta$pseudobulk_group, function(cells) {
  Matrix::rowSums(counts[, cells, drop = FALSE])
})
pseudobulk_counts <- do.call(cbind, pseudobulk_counts)

sample_meta <- data.frame(
  group = gsub("_.*", "", colnames(pseudobulk_counts)),
  organoid = gsub(".*_", "", colnames(pseudobulk_counts))
)
rownames(sample_meta) <- colnames(pseudobulk_counts)

sample_meta$group <- factor(sample_meta$group)
sample_meta$organoid <- factor(sample_meta$organoid)

dds <- DESeqDataSetFromMatrix(
  countData = pseudobulk_counts,
  colData = sample_meta,
  design = ~ organoid + group
)

dds <- DESeq(dds)

resultsNames(dds)
res <- results(dds, contrast = c("group", "0", "others")) 
res_df <- as.data.frame(res)
res_df$gene <- rownames(res_df)

# significance annotation
res_df <- res_df %>%
  mutate(
    sig = case_when(
      padj < 0.05 & log2FoldChange > 0.2 ~ "Up in clusters 0 + 3",
      padj < 0.05 & log2FoldChange < -0.2 ~ "Up in other clusters",
      TRUE ~ "Not significant"
    )
  )

volcano <- ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = sig)) +
  geom_point(alpha = 0.7, size = 1.5) +
  scale_color_manual(values = c(
    "Up in clusters 0 + 3" = "firebrick",
    "Up in other clusters" = "steelblue",
    "Not significant" = "gray80"
  )) +
  geom_vline(xintercept = c(-0.2, 0.2), linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  labs(
    x = expression(Log[2]~fold~change),
    y = expression(-Log[10]~adjusted~italic(P)),
    color = NULL,
    title = "Pseudobulk DE: clusters 0 + 3 vs others"
  ) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "top") +
  geom_text_repel(
    data = res_df %>% filter(padj < 0.01 & abs(log2FoldChange) > 1),
    aes(label = gene),
    #color = "black",
    size = 3,
    max.overlaps = 1000,
    box.padding = 0.5,
    point.padding = 0.3,
    segment.color = "gray40"
  )

ggsave("updated_06_24/pseudo_volcano2.pdf", volcano, width = 8, height = 6)
```

# Plot of features to explain the exhaustion signature

``` {r plot, echo = TRUE}
# genes
genes_of_interest <- c("Tcf7", "Emb", "Kit", "Klrg1", "Itgam", "Gzma") 

# grouping var
NK_and_ILC1$cluster.organoid <- paste0(NK_and_ILC1$NK_ILC1_clusters, "_", NK_and_ILC1$organoid)

# table for % expr per gene
freq_table <- lapply(genes_of_interest, function(gene) {
  FetchData(NK_and_ILC1, vars = c(gene, "NK_ILC1_clusters", "organoid")) %>%
    mutate(positive = !!sym(gene) > 0) %>%
    group_by(NK_ILC1_clusters, organoid) %>%
    summarise(freq = 100 * sum(positive) / n(), .groups = "drop") %>%
    mutate(Gene = gene)
}) %>% bind_rows()

colnames(freq_table) <- c("Cluster", "Organoid", "Frequency", "Gene")

frq_plot <- ggplot(freq_table, aes(x = Gene, y = Frequency, fill = Organoid)) +
  geom_col(position = position_dodge(width = 0.8), width = 0.7) +
  facet_wrap(~Cluster, nrow = 2) +
  scale_fill_manual(values = c("m95" = "lightblue", "m96" = "pink")) +
  labs(y = "% of Cells Expressing", x = "Gene", fill = "Organoid") +
  theme_minimal(base_size = 12) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("updated_06_24/expr_frequency.png", frq_plot, width = 10, height = 6)
```

# Feature plot (residency by org)

```{r res, echo = TRUE}
# midpoint
midpoint_val <- median(NK_and_ILC1@meta.data$Residency1, na.rm = TRUE)

Residency1_split_plot <- FeaturePlot(
  NK_and_ILC1,
  features = "Residency1",
  reduction = "umap",
  split.by = "organoid",     
  slot = "data"
) & 
  scale_color_gradient2(
    low = "purple",
    mid = "white",
    high = "red",
    midpoint = midpoint_val
  ) &
  theme_minimal() &
  ggtitle(paste0("Residency1 (mid=", round(midpoint_val, 3), ")"))

Residency1_split_plot
```

# Test of significance for tissue residency difference in organoids

```{r test, echo = TRUE}
# boxplot
residency_box <- ggplot(NK_and_ILC1@meta.data, aes(x = organoid, y = Residency1, fill = organoid)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, size = 1, alpha = 0.5) +
  labs(title = "Residency1 Score by Organoid", x = NULL, y = "Residency1 Module Score") +
  scale_fill_manual(values = c("m95" = "pink", "m96" = "lightblue")) +
  theme_minimal(base_size = 12)

# wilcoxon test
residency_test <- wilcox.test(Residency1 ~ organoid, data = NK_and_ILC1@meta.data)
residency_test$p.value

residency_box + 
  annotate("text", x = 1.5, y = max(NK_and_ILC1@meta.data$Residency1, na.rm = TRUE), 
           label = paste0("p = ", signif(residency_test$p.value, 3)), size = 4)


# pseudobulk + wicoxon
pseudobulk_scores <- NK_and_ILC1@meta.data %>%
  group_by(orig.ident, organoid) %>%
  summarise(mean_score = mean(Residency1, na.rm = TRUE)) %>%
  ungroup()

print(pseudobulk_scores)

wilcox_res <- wilcox.test(mean_score ~ organoid, data = pseudobulk_scores)
p_val <- signif(wilcox_res$p.value, digits = 3)

ggplot(pseudobulk_scores, aes(x = organoid, y = mean_score, fill = organoid)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.1, size = 3, alpha = 0.8) +
  theme_minimal(base_size = 13) +
  labs(
    y = "Residency1 (pseudobulk mean)",
    title = "Residency1 Pseudobulk Score by Organoid"
  ) +
  scale_fill_manual(values = c("m95" = "pink", "m96" = "lightblue")) +
  annotate("text", x = 1.5, y = max(pseudobulk_scores$mean_score) + 0.02,
           label = paste("p =", p_val), size = 5)


```
